### PR TITLE
Ventcrawl Overhaul

### PR DESCRIPTION
**Pull request by tigercat2000**
Port from -vg-: D3athrow/vgstation13#4004

Ventcrawling mobs now no longer get a teleportation menu, instead, they
are put inside the vents, the client gets an overlay of the pipes in the
pipe network, and it is up to them to navigate. Also adds alt-click ventcrawl and a few helper procs, because we were missing them.

Player Notes: Any ventcrawling mob now gets shoved into the vents and gets a way to navigate around, No more lolxeno's teleporting into the bridge- Everything that ventcrawls has to manually navigate now, and it takes time. Yes, this does mean pipes are somewhere things can hide, but they get ejected if the pipe is destroyed.

Coder Notes: Mostly self-contained, has a few hooks in atmos code out of necessity.

Important-ish note; With the differences between -vg-'s vent system and my relative inexperience with pipecode, this system can occasionally bug-out and just randomly put you in a different place than you expected. Remember to make posts on the issue tracker (if this gets merged of course)

Original Post;
>Feature freeze hasn't started yet!
Ventcrawling now actually puts you in the pipes and lets you move around.
You get an overlay of all the pipes in the network you're currently navigating.
Altclicking on a vent now specifically lets you crawl into that vent if there are other vents in the area.
You can now ventcrawl into scrubbers as well, because there's no real reason not to.

Fourm topic link; http://nanotrasen.se/phpBB3/viewtopic.php?f=48&t=4396

### CODE REVIEW

**Comment by Spacemanspark** [comment]
:+1:

**Comment by adrkiller59** [comment]
Can we have slime people ventcrawling back now?
If it's still too OP for you because "THEY KEEP ESCAPIN PRISOM" then put some special vents that block ventcrawling in the secure areas or something.

**Comment by Spacemanspark** [comment]
Not a bad idea, ADR.

**Comment by SamCroswell** [comment]
We already tried this system once, and it was (From what I remember) universally disliked.

**Comment by tigercat2000** [comment]
@SamCroswell This is not the same system. This is a brand new take on the idea. The old system is equivalent to the first supermatter we had- a beta idea.

**Comment by adrkiller59** [comment]
Of course it was disliked, it was like stumbling blindly in a labirynth, this is different, you at least get to see where you are.

**Comment by SamCroswell** [comment]
@tigercat2000 
Doesn't change the fundamental idea of the system. The only difference I see is you see the entire pipe layer instead of a 3x3 section of it.

**Comment by tigercat2000** [comment]
@SamCroswell You didn't see a 3x3 section of it with the original system, you saw literally nothing, just blindly navigating and spamming "BANK BING DONK" with sounds every time you hit a wall

**Comment by Finalsong** [comment]
I approve.

**Comment by tigercat2000** [comment]
Whoops, left in an unfinished project, remade the commit properly

**Comment by ComicIronic** [comment]
It isn't done, for the record. There's a bit extra I want to add to it, especially in terms of being able to weld scrubbers and a more OOP handling of which pipes you can move through.

**Comment by Fox-McCloud** [comment]
No. No. Nope. Nada. We tried this before; it was absolutely horrible. This may be an updated version of it, but it's essentially the same spirit of things.

:-1:

Was hoping this was the click on vents to ventcrawl system--guess not! Heh.

**Comment by tigercat2000** [comment]
@Fox-McCloud Is the premise really that bad? I seriously don't understand how literally a teleport button is better than this in any way. What is horrible about this?

**Comment by Fox-McCloud** [comment]
Contrary to preventing "lolxenos", it's going to hardened it even more--same way with lings; they'll just be able to hide in certain parts for indefinite periods of time without any real consequences. It's also going to lead to ventcrawlig--then the vents getting welded and getting stuck inside the tubes.

The helper procs for clicking on vents to ventcawl are great, but I don't think this would be a good system to implement on Para; it was tried before and nearly universally reviled.

**Comment by tigercat2000** [comment]
@Fox-McCloud That's forgetting that the actual gas inside the pipe affects them. Distro by default goes up to about 500KPA by 10 minutes into the round, that means that anything that isn't a xeno is going to die if it tries to hide inside the vents. Not to mention that wrenching up the pipe they are in/destroying it in any way causes them to fall out. It was tried before with an extremely buggy system for a very short time, I doubt xenos were even tried during that period. And all of that is not even slightly considering the fact that we were still a mediumRP server back then.

**Comment by Fox-McCloud** [comment]
If pressure impacts them, then this will wreck changeling's escape ability.

I distinctly remember a LOT of complaints from both Xeno and Changeling players from when it was implemented, even after the pipe overlay was tried out.

Sorry :-1:

**Comment by tigercat2000** [comment]
Pressure impacts them if they stay too long. Testing it shows that a monkey can survive in the pipes for around 1 minute before they get too damaged. A quick escape is still possible, you just can't immediately teleport across the entire map, you have to bother with looking for a vent in a quiet place and letting your health regen.

**Comment by Nomzosha** [comment]
Iunno how code works, but if hiding easily is such a big problem, why not make them get cramped if they stay for too long and give damage or someshit?

**Comment by Spacemanspark** [comment]
lings could use a small nerf. There is a reason that lesser form + vent crawling is a last resort. We should enforce this.

**Comment by veganzombeh** [comment]
In all honesty, this doesn't sound too bad. I don't really see why xenos and changelings hiding in vents is a particularly bad thing, since they're not really going to be able to achieve anything from inside. Really though, ventcrawling needs to not be instant. If not this, another option would be to select a destination, than slide through the vents like you do disposals pipes.

**Comment by fj45** [comment]
Really like this, while xenos hiding in the vents may be an issue...there should be ways to balance that and changlings seem much better balanced now.

**Comment by adrkiller59** [comment]
How to solve the alien infestation in the atmos pipes trouble: make the fireaxe be able to destroy atmos pipes when wielded. 
Oh god that would make Buck OP, nevermind, forget it.

Oh what about t-ray scanners being able to show creatures in pipes? Makes sense aye?

**Comment by tigercat2000** [comment]
>ComicIronic commented 5 hours ago
>>Can you make it so people with thermals can see inside the vents?

>Not easily, since I'm pretty sure the thermal view effect is a Byond feature using SEE_MOBS.

**Comment by fj45** [comment]
T-rays being able to see mobs in pipes would be good, but in no way a proper way to balance them, T-rays are crap.

**Comment by TheDZD** [comment]
I think Tiger is trying to point out that there's no easy way to make mobs inside of pipes visible short of completely changing how vision works.

**Comment by tigercat2000** [comment]
Already done, from vg's source- Tigercat2000/Paradise@b96f5ea7efb1b3718eccf0867ffff05b8a7a8bec

**Comment by fj45** [comment]
Err, I should ask, have you added ventcrawling/welding to scrubbers? Because frankly there doesn't need to be twice as many things to seal.

**Comment by tigercat2000** [comment]
@fj45 yes, this does add ventcrawling and welding to scrubbers, because there isn't really any reason not to have it. The waste loop is not very habitable.

**Comment by fj45** [comment]
If atmos is even half competent, its empty. Even on pretty crazy rounds the waste loop is nearly always empty.

And I did give a reason, it doubles the workload to seal an area (already something that is very difficult to properly accomplish.)

To turn that around, WHY do we need scrubbercrawling?

**Comment by tigercat2000** [comment]
That's the point, it's either empty or on fire, it's not very safe to enter, either you get burned to death or low pressure

And scrubbercrawling is for consistency, it doesn't seem like that big of a deal. After-all, if you want to seal an area, you probably want to seal the scrubbers regardless, or deconstruct them, to be safe against air alarm tampering and the such.

**Comment by FalseIncarnate** [comment]
Personally, I heavily support this.

Yes, it encourages the xenos to hide in pipes, which they pretty much do already in the Aliens universe. They aren't unstoppable juggernauts who charge into battle head-on, but rather ambushing predators that utilize their surroundings to isolate and entrap their prey.

Changelings are able to use the pipes still, but the pressure prevents them from indefinitely hiding as they will die. Perhaps a slight nerf in some regards, sure, but it gives the ling a sense of urgency to move through the pipes quickly rather than dawdle all shift. If the amount of time they can survive in the pipe is too limited, you could always make tweaks to lesser form to give them a bit more time through some chems to offset the damage (but not negate it).

As for scrubber crawling, I wholeheartedly am supportive of this. Currently, the only external vents (ones to space) are connected to scrubbers, not vents, so space-bound xenos were forced to actually breach the station rather than slip on board quietly. It does make sealing an area more difficult, but being able to weld a scrubber also could prevent the AI from panic siphoning you on a malf round, or perhaps even prevent the scrubber back-up random event from spewing acid into your face (that might not be included right now).

TL;DR: :+1:

**Comment by ZomgPonies** [comment]
I'm willing to :+1: give it a shot again. It sure as fuck beats the lousy insta-jump menus.

**Comment by ZomgPonies** [comment]
Seems so far it's 10ish for, 2 against?

Although since it's been tested in a more primitive form in the past and been reverted, perhaps it should get tagged controversial and a forum post/vote be added for a bigger voting sample.

**Comment by ZomgPonies** [comment]
Yeah go put up a forum post with a vote and get more community input please.

**Comment by tigercat2000** [comment]
Done. http://nanotrasen.se/phpBB3/viewtopic.php?f=48&t=4396

**Comment by Artorp** [comment]
:+1: I'd like to see how this plays out, could be obnoxious, could make everything better. We'll see.

**Comment by CrAzYPiLoT-SS13** [comment]
I really support this. Why have some boxes when you can have a totally realistic mechanic?

**Comment by ZomgPonies** [comment]
Seems pretty clear cut according to the forums. I'll merge this in. We'll test it out, and if it's really awful then look at our options.
